Source/Core/DolphinQt/Debugger/MemoryWidget.cpp

```diff
@@ -476,12 +495,123 @@ void MemoryWidget::OnSearchAddress()
   m_search_offset->setPalette(offset_palette);
 }
 
-void MemoryWidget::ValidateSearchValue()
+void MemoryWidget::ValidateAndPreviewInputValue()
 {
+  m_data_preview->clear();
+  QString input_text = m_data_edit->text();
+
+  if (input_text.isEmpty())
+    return;
+
+  // Remove any spaces
+  if (!m_input_ascii->isChecked())
+    input_text = input_text.replace(QLatin1Char(' '), QStringLiteral(""));
+
   QFont font;
   QPalette palette;
+  QString hex_string;
+  int padding = 8;
+  int hex = 16;
+  bool good = true;
 
-  if (!IsValueValid())
+  if (m_input_ascii->isChecked())
+  {
+    // MemoryView uses Latin1.
+    QByteArray bytes = input_text.toLatin1();
+
+    // An unrestrained size is fine, but it needs to truncate in the preview.
+    if (bytes.size() > 8)
+    {
+      bytes.truncate(8);
+      hex_string = QString::fromLatin1(bytes.toHex()).append(QStringLiteral("...."));
+    }
+    else
+    {
+      hex_string = QString::fromLatin1(bytes.toHex());
+    }
+  }
+  else if (m_input_float->isChecked())
+  {
+    const float float_in = input_text.toFloat(&good);
+
+    if (good)
+    {
+      const u32 hex_out = Common::BitCast<u32>(float_in);
+      hex_string = QStringLiteral("%1").arg(hex_out, padding, hex, QLatin1Char('0')).toUpper();
+    }
+  }
+  else if (m_input_integer->isChecked())
```

Comment:
I really think the logic is overkill. What about using a dropdown menu with the proper input type (u8, s8, u16, s16, u32, u64, float, double, hex, ascii, ...)?

Moreover, I think the `m_type_*` members are misused. IIRC, they were only used to specified the representation of the memory view, not the input type. This use would be counterintuitive.

Reply:
6 different options for something that can be done with 1 option and some logic seems like overkill to me.  Is it really unreasonable to expect someone to choose u8 view if they want to input a u8? The view doesn't determine type, only padding/length.

Reply:
I find this actually pretty intuitive for integers and floats, though ymmv of course. For a hex string, less so, maybe that one should never be padded?

Reply:
> 6 different options for something that can be done with 1 option and some logic seems like overkill to me.

"Can be done with 1 option" and tons of branching is IMHO clearly overkill. On top of that, checking the first character each time to see if its signed or not is redundant and way harder to read.

The same goes for the padding heuristic which is everywhere. Whereas knowing the exact type gives you the appropriate padding without having to compute it.

> Is it really unreasonable to expect someone to choose u8 view if they want to input a u8? The view doesn't determine type, only padding/length ( I guess you could call that type tho).

Yes it is, if it was not for the hex string preview, I don't think many people would have understood it. Especially, since there is an ASCII type and a float type in both group. So that's extremely confusing.

Reply:
How would you turn -20 into an 8 bit hexadecimal and a 16 bit hexadecimal QString? I'm more than happy to improve upon it, with a bit of guidance.

My basic goal is adding integer + float. Allowing the option to input 1 -> 00000001.  Add a spaced out preview to easily see what you're inputting and how many bytes it is.

Should input and view be combined into one thing? Is there any time that'd be annoying?

Reply:
> How would you turn -20 from a QString into an 8 bit hexadecimal and a 16 bit hexadecimal QString? I'm more than happy to improve upon it, with a bit of guidance.

A `QComboBox` can be used with the proper input type (u8, s8, u16, s16, u32, s32, u64, s64, float, double, hex, ascii, ...). Sticking to radio button sounds like a bad design choice to me. With the `QComboBox` index, the validation/preview can be simplified and will boil down to a switch and an enum class [like this](https://github.com/dolphin-emu/dolphin/blob/0204b119e904d358ba7084d18a84f77929a5f7e5/Source/Core/DolphinQt/Debugger/NetworkWidget.cpp#L392-L430) and picking the right QString method `(toShort, toInt, ...)`.

For instance:
```c++
switch (combo_id)
  {
  case InputComboId::S16:
  {
    short value = input_text.toShort(&good, 10);
    if (good)
      hex_string = QStringLiteral("%1").arg(value, 4, 16, QLatin1Char('0')).toUpper();
    break;
  }
  case InputComboId::S32:
  {
    int value = input_text.toInt(&good, 10);
    if (good)
      hex_string = QStringLiteral("%1").arg(value, 8, 16, QLatin1Char('0')).toUpper();
    break;
  }
  // Etc. (...)
  }
```

> Should input and view be combined into one thing? Is there any time that'd be annoying?
I could also make it so if input size > intended size it defaults to string of bytes instead of continued padding.

I think changing the view behaviour is out of scope of the PR.

Reply:
That code doesn't give the correct output and the fix is the same padding convolution that's trying to be avoided.

Reply:
I thought that it would have behaved like libfmt's format or printf but I was wrong.

Here is an alternative using libfmt sprintf:
```c++
  case InputComboId::S16:
  {
    short value = input_text.toShort(&good, 10);
    if (good)
      hex_string = QString::fromStdString(fmt::sprintf("%04hX", value));
    break;
  }
```

Using `fmt::format` follows python's format and doesn't print the intended number:
```c++
// Code testing
short value = -42;
qDebug() << QStringLiteral("%1").arg(value, 4, 16, QLatin1Char('0')).toUpper();
qDebug() << QString::fromStdString(fmt::format("{:04X}", value));
// You'll need to use printf specifiers and length modifiers here
qDebug() << QString::fromStdString(fmt::sprintf("%04hX", value));
```

Reply:
Thank you! I knew fmt was incorrect, but didn't know about fmt sprintf.

Reply:
I didn't find a simpler method for float than BitCast, but everything else seems improved.

One problem I have with making 12 different input types, is all the red you will see if you go out of range. With just the integer type, at least you see an output and can adjust it rather than just getting hit with an invalid s16 input that's out of range.